File: BOJ/python3/3067.py
'''
Coins
https://www.acmicpc.net/problem/3067
'''
import sys
import time
start = time.time()  # 시작 시간
sys.stdin = open("input.txt", "r")  # input.txt 파일로 입력


T = int(input())
for _ in range(T):
    N = int(input())
    C = [0] + list(map(int, input().split()))
    M = int(input())

    dp = [0]*(M+1)
    dp[0] = 1
    for n in range(1, N+1):
        for m in range(C[n], M+1):
            dp[m] += dp[m-C[n]]

    ans = dp[M]
    print(ans)

# 현재시각 - 시작시간 = 실행 시간(sec)
print("--------------------")
print(f"Elapsed Time: {round(time.time() - start, 3)}s")

File: BOJ/nodejs/3067.js
/*
Coins
boj.kr/3067
 */
const start = new Date().getTime();

const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().split("\n");
//const input = fs.readFileSync("/dev/stdin").toString().split("\n");

let ln = 0; // line number
const T = 1 * input[ln++];
for (let t = 0; t < T; t++) {
  const N = 1 * input[ln++];
  const coins = input[ln++].split(" ").map(Number);
  const M = 1 * input[ln++];

  let ans;
  const dp = new Array(M + 1).fill(0); // 금액 M 일 때, 경우의 수
  dp[0] = 1; // 금액 0 일 때, 경우의수 1 (공집합)
  for (let n = 0; n < N; n++) {
    let coin = coins[n]; // 동전의 가치
    for (let m = coin; m < M + 1; m++) {
      let price = m; // 금액 m
      dp[price] += dp[price - coin];
    }
  }

  ans = dp[M];
  console.log(ans);
}

const end = new Date().getTime();
console.log("--------------------");
console.log(`Elapsed Time: ${(new Date().getTime() - start) / 1000}s`);
